plugins/protocols/session-lock.cpp

```diff
@@ -0,0 +1,198 @@
+        {
+          public:
+            lock_surface_keyboard_interaction(wlr_surface *surface) : surface(surface)
+            {}
+
+            void handle_keyboard_enter(wf::seat_t *seat)
+            {
+                // TODO: use a keyboard grab instead?
+                wlr_seat_keyboard_enter(seat->seat, surface, nullptr, 0, nullptr);
+            }
+
+            void handle_keyboard_leave(wf::seat_t *seat)
+            {
+                // TODO: refocus?
+                wlr_seat_keyboard_clear_focus(seat->seat);
+            }
+
+            void handle_keyboard_key(wf::seat_t *seat, wlr_keyboard_key_event event)
+            {
+                wlr_seat_keyboard_notify_key(seat->seat, event.time_msec, event.keycode, event.state);
+            }
+
+          private:
+            wlr_surface *surface;
+        };
+
+        wf::keyboard_interaction_t& keyboard_interaction()
+        {
+            return *interaction;
+        }
+
+      private:
+        wf::keyboard_interaction_t *interaction;
+    };
+
+    class wayfire_session_lock
+    {
+      public:
+        wayfire_session_lock(wf_session_lock_impl *parent, wlr_session_lock_v1 *data)
+        {
+            plugin = parent;
+            lock   = data;
+
+            new_surface.set_callback([this] (void *data)
+            {
+                wlr_session_lock_surface_v1 *lock_surface = (wlr_session_lock_surface_v1*)data;
+                uint32_t serial = wlr_session_lock_surface_v1_configure(lock_surface,
+                    lock_surface->output->width,
+                    lock_surface->output->height);
+                LOGI("surface_configure output=", lock_surface->output->name, " ",
+                    lock_surface->output->width, "x",
+                    lock_surface->output->height,
+                    " serial=", serial);
+
+                surfaces.push_back(lock_surface);
+
+                auto node = new std::shared_ptr<wf::scene::wlr_surface_node_t>(
+                    new lock_surface_node_t(lock_surface->surface, true));
+
+                lock_surface->data = node;
+                auto target_layer  = wf::scene::layer::TOP;
+                auto wo = wf::get_core().output_layout->find_output(lock_surface->output);
+                auto layer_node = wo->node_for_layer(target_layer);
+
+                (*node)->set_enabled(true);
+                wf::scene::add_front(layer_node, *node);
+                wf::get_core().seat->set_active_node(*node);
+
+                lock_surface_destroy.set_callback([lock_surface, this] (void *data)
+                {
+                    auto node = (std::shared_ptr<wf::scene::wlr_surface_node_t>*)lock_surface->data;
+                    if (node != nullptr)
+                    {
+                        wf::scene::remove_child(*node);
+                    }
+
+                    lock_surface_destroy.disconnect();
+                });
+                lock_surface_destroy.connect(&lock_surface->events.destroy);
```

Comment:
You will need one listener per created surface (because listeners cannot connect to multiple object's signals unlike Wayfire's signals). So lock_surface_destroy() probably will have to be moved to the node class.

Reply:
Thanks for the suggestion. Done, I think.